enhancements/network/egress-qos.md

```diff
@@ -0,0 +1,187 @@
+
+```yaml
+kind: EgressQoS
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: default
+  namespace: default
+spec:
+  egress:
+  - dscp: 46
+    dstCIDR: 0.0.0.0/0
+  - dscp: 30
+    dstCIDR: 1.2.3.4/32
+```
+the equivalent of:
+```bash
+ovn-nbctl qos-add node1 from-lport 1 "ip4.src == <default_ns_address_set> && ip4.dst == 0.0.0.0/0" dscp=46
+ovn-nbctl qos-add node1 from-lport 2 "ip4.src == <default_ns_address_set> && ip4.dst == 1.2.3.4/32" dscp=30
+```
+will be executed.
+
+In addition it'll watch pods to decide if further updates are needed, for example:
+when another pod `app2` comes up in the namespace on node `node2`, the controller will
+attach the existing `QoS` object to the local switch of `node2`.
+
+IPv6 will also be supported, given the following `EgressQoS`:
+```yaml
+apiVersion: k8s.ovn.org/v1
+kind: EgressQoS
+metadata:
+  name: default
+  namespace: default
+spec:
+  egress:
+  - dscp: 48
+    dstCIDR: 2001:0db8:85a3:0000:0000:8a2e:0370:7330/124
+```
+and a single pod with the IP `fd00:10:244:2::3` in the namespace, the controller will create the relevant QoS object that will result in a similar flow to this on the pod's node:
+```bash
+ cookie=0x6d99cb18, duration=63.310s, table=18, n_packets=0, n_bytes=0, idle_age=63, priority=555,ipv6,metadata=0x4,ipv6_src=fd00:10:244:2::3,ipv6_dst=2001:db8:85a3::8a2e:370:7330/124 actions=mod_nw_tos:192,resubmit(,19)
+```
+
+### User Stories
+#### Story 1
+
+As a user of OpenShift, I should be able to mark egress traffic coming from a specific namespace with a valid DSCP value.
+
+### API Extensions
+
+A new namespace-scoped CRD is introduced:
+
+```go
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=egressqoses
+// +kubebuilder::singular=egressqos
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// EgressQoS is a CRD that allows the user to define a DSCP value
+// for pods egress traffic on its namespace to specified CIDRs.
+// Traffic from these pods will be checked against each EgressQoSRule in
+// the namespace's EgressQoS, and if there is a match the traffic is marked
+// with the relevant DSCP value.
+type EgressQoS struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   EgressQoSSpec   `json:"spec,omitempty"`
+	Status EgressQoSStatus `json:"status,omitempty"`
+}
+
+// EgressQoSSpec defines the desired state of EgressQoS
+type EgressQoSSpec struct {
+	// a collection of Egress QoS rule objects
+	Egress []EgressQoSRule `json:"egress"`
+}
+
+type EgressQoSRule struct {
+	// Dscp marking value for matching pods' traffic.
+	// +kubebuilder:validation:Maximum:=63
+	// +kubebuilder:validation:Minimum:=0
+	Dscp int `json:"dscp"`
+
+	// DstCIDR specifies the destination's CIDR. Only traffic heading to this CIDR will be marked with DSCP.
+	DstCIDR string `json:"dstCIDR"`
```

Comment:
Should this be optional? In other words, if the user wants all egress traffic marked with the same QoS value, there should be no need to specify 0.0.0.0/0

Reply:
Need kubebuilder validation for addresses and valid dscp code values.

Reply:
> Should this be optional? In other words, if the user wants all egress traffic marked with the same QoS value, there should be no need to specify 0.0.0.0/0

I don't have a strong opinion about this, do you think defaulting to 0.0.0.0/0 or ::/0 is clearer?



> Need kubebuilder validation for addresses and valid dscp code values.

For the dscp I added one, for addresses I'm a bit hesitant about writing a good enough regex that covers all IPv4/6 CIDRs. In any case, we'll parse the addresses with `net.ParseCIDR`.

Reply:
Check how the service CIDR and cluster IP subnets are validated in network.config.openshift.io object in `openshift/api`.

Reply:
If the 80% users want all traffic in the namespace to be marked a certain way, then the dst cidr field seems like an overhead. OTOH, if specifying only a subset of destinations for which the packets should be marked with a certain DSCP value is the more common case, then I would lean on keeping the field mandatory.

Reply:
after also talking with Tim and Dan this field should be optional, updated.
thanks :)